File: nmon-logger-syslog-ng/etc/nmon-logger/bin/fifo_reader.py
#!/usr/bin/env python

import os
import sys
import optparse
import logging
import re
import subprocess

# Releases Notes:

# - March 2017, V1.0.0: Guilhem Marchand, Initial version
# - 2017/04/09, V1.0.01: Guilhem Marchand, mirror update from TA-nmon

# script version
version = '1.0.01'

#################################################
#      Variables
#################################################

# Set logging format
logging.root
logging.root.setLevel(logging.DEBUG)
formatter = logging.Formatter('%(levelname)s %(message)s')
handler = logging.StreamHandler()
handler.setFormatter(formatter)
logging.root.addHandler(handler)

# APP_VAR directory
APP_VAR = '/var/log/nmon/var'
if not os.path.exists(APP_VAR):
    logging.info(
        'The application var directory does not exist yet, we are not ready to start')
    sys.exit(0)

# APP
APP = "/etc/nmon-logger"

# Verify APP exist
if not os.path.exists(APP):
    msg = 'The Application root directory could not be found, is the TA-nmon ? We tried: ' + \
          str(APP)
    logging.error(msg)
    sys.exit(1)

# fifo_reader.sh
fifo_reader = APP + "/bin/fifo_reader.sh"

#################################################
#      Arguments
#################################################

parser = optparse.OptionParser(usage='usage: %prog [options]', version='%prog '+version)

parser.add_option('-F', '--fifo', action='store', type='string', dest='fifo_name',
                  help='set the fifo file to be read')
parser.add_option('--dumpargs', action='store_true', dest='dumpargs',
                  help='only dump the passed arguments and exit (for debugging purposes only)')

(options, args) = parser.parse_args()

if options.dumpargs:
    print("options: ", options)
    print("args: ", args)
    sys.exit(0)

if not options.fifo_name:
    logging.error(
        'The fifo file option has not been set (-F fifo_name or --fifo fifo_name)')
    sys.exit(1)
else:
    fifo_name = options.fifo_name

# define the full path to the fifo file
fifo_path = APP_VAR + '/nmon_repository/' + fifo_name + '/nmon.fifo'

# At startup, rotate any existing non empty .dat file if nmon_data.dat is not empty

# define the various files to be written

# realtime files
nmon_config_dat = APP_VAR + '/nmon_repository/' + fifo_name + '/nmon_config.dat'
nmon_header_dat = APP_VAR + '/nmon_repository/' + fifo_name + '/nmon_header.dat'
nmon_data_dat = APP_VAR + '/nmon_repository/' + fifo_name + '/nmon_data.dat'
nmon_timestamp_dat = APP_VAR + '/nmon_repository/' + fifo_name + '/nmon_timestamp.dat'
nmon_external_dat = APP_VAR + '/nmon_repository/' + fifo_name + '/nmon_external.dat'
nmon_dat = {nmon_config_dat, nmon_header_dat, nmon_timestamp_dat, nmon_data_dat, nmon_external_dat}

# Manage existing files and do the rotation if required
if os.path.exists(nmon_data_dat) and os.path.getsize(nmon_data_dat) > 0:
    for file in nmon_dat:
        rotated_file = str(file) + ".rotated"
        if os.path.isfile(rotated_file):
            os.remove(rotated_file)
        if os.path.isfile(file):
            os.rename(file, rotated_file)

elif os.path.exists(nmon_data_dat):
    for file in nmon_dat:
        if os.path.isfile(file):
            os.remove(file)

####################################################################
#           Main Program
####################################################################

# Verify the fifo file exists, and start processing
if not os.path.exists(fifo_path):
    logging.info(
        'The fifo file ' + fifo_path + ' does not exist yet, we are not ready to start')
    sys.exit(0)
else:
    # we use the fifo_reader.sh to read the fifo file, benchmarks have shown more stability than
    # opening the fifo file in pure Python
    cmd = fifo_reader + " " + fifo_path
    p = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)

    while 1:
        line = p.stdout.readline()
        if line == '' and p.poll() != None:
            break

        # Manage nmon config
        nmon_config_match = re.match(r'^[AAA|BBB].+', line)
        nmon_header_match = re.match(r'^(?!AAA|BBB|TOP)[a-zA-Z0-9\-\_]*,(?!T\d{3,}).*', line)
        nmon_header_TOP_match = re.match(r'^TOP,(?!\d*,)', line)
        nmon_timestamp_match = re.match(r'^ZZZZ,T\d*', line)

        if nmon_config_match:
            with open(nmon_config_dat, "ab") as nmon_config:
                nmon_config.write(line)

        elif nmon_header_match:
            with open(nmon_header_dat, "ab") as nmon_header:
                nmon_header.write(line)

        elif nmon_header_TOP_match:
            with open(nmon_header_dat, "ab") as nmon_header:
                nmon_header.write(line)

        # timestamp management: write the nmon timestamp in nmon_data and as well nmon_timestamp for later use
        elif nmon_timestamp_match:
            with open(nmon_timestamp_dat, "ab") as nmon_timestamp:
                nmon_timestamp.write(line)
            with open(nmon_data_dat, "ab") as nmon_data:
                nmon_data.write(line)

        else:
            with open(nmon_data_dat, "ab") as nmon_data:
                nmon_data.write(line)

File: nmon-logger-splunk-hec/etc/nmon-logger/bin/fifo_reader.pl
#!/usr/bin/perl

# Program name: fifo_reader.pl
# Compatibility: Perl x
# Purpose - read nmon data from fifo file
# Author - Guilhem Marchand
# Date of first publication - March 2017

# Releases Notes:

# - March 2017, V1.0.0: Guilhem Marchand, Initial version
# - 2017/04/01, V1.0.1: Guilhem Marchand, Update path discovery
# - 2017/06/05, V1.0.2: Guilhem Marchand:
#                                          - Mirror update of the TA-nmon

$version = "1.0.2";

use Getopt::Long;
use File::stat;
use File::Copy;

#################################################
##      Arguments Parser
#################################################

# Default values
my $APP       = "";
my $fifo_name = "";

$result = GetOptions(
    "fifo=s"  => \$fifo_name,    # string
    "version" => \$VERSION,      # flag
    "help"    => \$help          # flag
);

# Show version
if ($VERSION) {
    print("fifo_reader.pl version $version \n");

    exit 0;
}

# Show help
if ($help) {

    print( "

Help for fifo_reader.pl:

The script should be run in the backgroud to continously read nmon data from fifo files.

Available options are:

--fifo <name of fifo> :Name of the pre-configured fifo file
--version :Show current program version \n
"
    );

    exit 0;
}

# main APP
my $APP = "/etc/nmon-logger";

# Verify existence of APP
if ( !-d "$APP" ) {
    print(
"\n$time ERROR: The Application root directory could not be found, is the TA-nmon installed ?\n"
    );
    die;
}

# var directory
my $APP_VAR     = "/var/log/nmon-logger/var";

if ( !-d "$APP_VAR" ) {
    print(
"\n$time INFO: main var directory not found ($APP_VAR),  no need to run.\n"
    );
    exit 0;
}

# check fifo_name
if ( not "$fifo_name" ) {
    print("\n$time ERROR: the --fifo_name <name of fifo> is mandatory\n");
    die;
}

# define the full path to the fifo file
my $fifo_path = "$APP_VAR/nmon_repository/$fifo_name/nmon.fifo";

# At startup, rotate any existing non empty .dat file if nmon_data.dat is not empty

# define the various files to be written

# realtime files
my $nmon_config_dat = "$APP_VAR/nmon_repository/$fifo_name/nmon_config.dat";
my $nmon_header_dat = "$APP_VAR/nmon_repository/$fifo_name/nmon_header.dat";
my $nmon_data_dat   = "$APP_VAR/nmon_repository/$fifo_name/nmon_data.dat";
my $nmon_external_dat   = "$APP_VAR/nmon_repository/$fifo_name/nmon_external.dat";
my $nmon_timestamp_dat =
  "$APP_VAR/nmon_repository/$fifo_name/nmon_timestamp.dat";

@nmon_dat = (
    "$nmon_config_dat", "$nmon_header_dat",
    "$nmon_data_dat",   "$nmon_timestamp_dat",
    "$nmon_external_dat"
);

# Remove any existing rotated file
foreach $file (@nmon_dat) {
    $rotated_file = "$file.rotated";
    if (-e $rotated_file) {
        unlink $rotated_file;
    }
}

# Manage existing files and do the rotation if required
if ( !-z $nmon_data_dat ) {
    foreach $file (@nmon_dat) {
        $rotated_file = "$file.rotated";
        move( $file, $rotated_file );
    }
}
else {
    foreach $file (@nmon_dat) {
        if ( -e $file ) {
            unlink $file;
        }
    }
}

####################################################################
#############		Main Program
####################################################################

if ( !-p $fifo_path ) {
    print(
"\n$time INFO: The fifo file $fifo_path does not exist yet, we are not ready to start.\n"
    );
    exit 0;

}
else {

    my $fifoh;

# Open the named pipe "a la shell" to ensure that we we will quite when the nmon process has ended as well
    open( $fifoh, "$APP/bin/fifo_reader.sh $fifo_path|" );

        while (<$fifoh>) {
            chomp($_);

            $nmon_config_match     = '^[AAA|BBB].+';
            $nmon_header_match     = '^(?!AAA|BBB|TOP)[a-zA-Z0-9\-\_]*,(?!T\d{3,})[^,]*,(?!T\d{3,})[^,]*.*';
            $nmon_header_TOP_match = '^TOP,(?!\d*,)';
            $nmon_timestamp_match  = '^ZZZZ,T\d*';

            if ( $_ =~ /$nmon_config_match/ ) {
                open( my $fh, '>>', $nmon_config_dat )
                  or die "Could not open file '$nmon_config_dat' $!";
                print $fh "$_\n";
                close $fh;
            }

        elsif ( $_ =~ /$nmon_header_match/ ) {
            open( my $fh, '>>', $nmon_header_dat )
              or die "Could not open file '$nmon_header_dat' $!";
            print $fh "$_\n";
            close $fh;
        }

        elsif ( $_ =~ /$nmon_header_TOP_match/ ) {
            open( my $fh, '>>', $nmon_header_dat )
              or die "Could not open file '$nmon_header_dat' $!";
            print $fh "$_\n";
            close $fh;
        }

        elsif ( $_ =~ /$nmon_timestamp_match/ ) {
            open( my $fh, '>>', $nmon_timestamp_dat )
              or die "Could not open file '$nmon_timestamp_dat' $!";
            print $fh "$_\n";
            close $fh;
            open( my $fh, '>>', $nmon_data_dat )
              or die "Could not open file '$nmon_data_dat' $!";
            print $fh "$_\n";
            close $fh;
        }

        else {
            open( my $fh, '>>', $nmon_data_dat )
              or die "Could not open file '$nmon_data_dat' $!";
            print $fh "$_\n";
            close $fh;
        }

    }
    close $fifoh;
    exit(0);

}
